rust/tools/authorization-logic/src/souffle/lowering_ast_datalog.rs

```diff
@@ -0,0 +1,341 @@
+/// an extra argument as above).
+///
+///  Next we need to translate delegations which should pass beliefs from one
+///  principal to another. For example, if we have
+///      A says B canSay fact(x) :- fact1(x)
+///  we do the usual translation resulting in
+///      (1) A says B canSay fact(x) :- A says fact1(x)
+///  but we also add the additional rule
+///      (2) A says fact(x) :- x says fact(x), A says x canSay fact(x)
+///  where x is a fresh variable.
+///  The result is that rule (1) gives the condition for delegation
+///  and rule (2) passes beliefs from B to A when the condition from (1) is met.
+///
+/// Finally we need to translate canActAs which should pass properties from
+/// one principal to another:
+/// Assertions of either the form
+///      A says B canActAs C :- ...
+/// or
+///      A says B <anyPredicate(...)> :-
+/// are claims by A that B has a property (where canActAs is just a special
+/// case). Syntactically these are both treated as "A says B <verbphrase>" with
+/// different instances of <verbphrase>. Whenever some other principal can act
+/// as B, we want to pass these properties from B to that other principal.
+/// (For the special case where this property is B canActAs C, this essentially
+/// makes canActAs transitive). In either case, we add an additional rule
+///    A says x verbphrase <- A says x canActAs B, A says B verbphrase
+/// which will add "verbphrase" as a property of D whenever we can prove
+/// the assertion "A says D canActAsB".
+///
+/// ## Translation of Queries
+/// Neither Souffle nor SecPal support queries, but this language does.
+/// Queries check of a single unconditional assertion is true. A query of
+/// the form:
+///
+///      Q_NAME = query ASSERTION(<ARGS>) ?
+///
+/// is translated into an assertion with just one argument, and this
+/// assertion is made an output, so we can simply check the CSV that
+/// souffle emits as in:
+///
+///      Q_NAME("dummy_var") :- grounded("dummy_var"), ASSERTION(<ARGS>).
+///      grounded("dummy_var").
+///      .output Q_NAME
+
+// Note that this puts args_ on the front of the list of arguments because
+// this is the conveninet way for it to work in the contexts in which it
+// is used.
+fn push_onto_pred(modifier: String, mut args_: Vec<String>, pred: &AstPredicate) -> AstPredicate {
+    let new_name = modifier + &pred.name.clone();
+    for a in &pred.args {
+        args_.push(a.to_string());
+    }
+    AstPredicate {
+        name: new_name,
+        args: args_.to_vec(),
+    }
+}
+
+fn push_prin(modifier: String, p: &AstPrincipal, pred: &AstPredicate) -> AstPredicate {
+    let mut args = Vec::new();
+    args.push(p.name.clone());
+    push_onto_pred(modifier, args, pred)
+}
+
+// This struct only contains a counter for generating new
+// fresh variables.
+pub struct LoweringToDatalogPass {
+    fresh_var_count: u32,
+}
+
+impl LoweringToDatalogPass {
+    // This is the only public function in the struct.
```

Comment:
I am not sure this comment is useful. I think it should either be converted to a doc comment explaining what it does, or removed.

Reply:
Fixed.